### PR TITLE
Policy audit failure scenarios

### LINKED ISSUE #1987: Rspec Unit Tests are Present for Existing Audit Events
Unit tests are currently missing for few audit events to verify they render the correct log output for the given inputs.

### PR DESCRIPTION
**Pull request by egvili**
### What does this PR do?
Add audit logs for failure scenarios involving loading a policy.

### What ticket does this PR close?
Resolves #1987 

#### Change log
- [ ] The CHANGELOG has been updated, or
- [X] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [X] This PR includes new unit and integration tests to go with the code changes, or
- [ ] The changes in this PR do not require tests

#### Documentation
- [X] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs - [doc issue](https://github.com/cyberark/conjur-docs/issues/1413)
- [ ] This PR does not require updating any documentation

### CODE REVIEW

**Review by uCatu — APPROVED** [review]

File: app/controllers/policies_controller.rb [hunk 1]
```diff
@@ -67,6 +79,22 @@ def find_or_create_root_policy
 
   private
 
+  def audit_success(policy_version)
```
Comment (orenbm) [inline app/controllers/policies_controller.rb:82]: the param here is `policy_version` bu all the calling methods use `policy`. can we align this?

File: app/controllers/policies_controller.rb [hunk 2]
```diff
@@ -67,6 +79,22 @@ def find_or_create_root_policy
...
+    policy_version.policy_log.lazy.map(&:to_audit_event).each do |event|
+      Audit.logger.log(event)
+    end
+  end
+
+  def audit_failure(err, operation)
+    Audit.logger.log(Audit::Event::Policy.new(
```
Comment (orenbm) [inline app/controllers/policies_controller.rb:89]: i would move `Audit::Event::Policy.new(` to a new line

File: spec/models/audit/event/policy_spec.rb [hunk 3]
```diff
@@ -19,12 +19,14 @@
       role: user
     )
   end
+  let(:error_message) { nil }
```
Comment (orenbm) [inline spec/models/audit/event/policy_spec.rb:22]: please add an empty line before this

**Review by orenbm — CHANGES_REQUESTED** [review]
left 2 comments.

please also address the CC comments.

File: app/controllers/policies_controller.rb [hunk 4]
```diff
@@ -67,6 +40,41 @@ def find_or_create_root_policy
 
   private
 
+  def load_policy(privilege, loader_class, delete_permitted)
```
Comment (orenbm) [inline app/controllers/policies_controller.rb:43]: i would change `privilege` to `action`

File: app/models/audit/event/policy.rb [hunk 5]
```diff
@@ -38,23 +40,34 @@ def to_s
       # TODO: See issue https://github.com/cyberark/conjur/issues/1608
       # :reek:NilCheck
       def message
-        past_tense_verb = @operation.to_s.chomp('e') + "ed"
-        "#{user&.id} #{past_tense_verb} #{@subject}"
+        attempted_action.message(
+          success_msg: success_message,
+          failure_msg: "Failed loading policy",
```
Comment (orenbm) [inline app/models/audit/event/policy.rb:45]: @shulifink can you review this message please?

**Review by orenbm — COMMENTED** [review]
left a small comment.
please also get a review from @shulifink

File: app/models/audit/event/policy.rb (hunk 5 above)
Comment (shulifink) [inline app/models/audit/event/policy.rb:45]:
```suggestion
          failure_msg: "Failed to load policy",
```

**Review by shulifink — CHANGES_REQUESTED** [review]
one tiny change

File: cucumber/api/features/policy_load_errors.feature [hunk 6]
```diff
@@ -82,3 +95,10 @@ Feature: Policy loading error messages
       }
     }
     """
+    And there is an audit record matching:
+    """
+      <85>1 * * conjur * policy
+      [auth@43868 user="cucumber:user:admin"][subject@43868]
+      [action@43868 result="failure" operation="add"]
+      [meta sequenceId="4"] Failed loading policy: policy_text is not present
```
Comment (shulifink) [inline cucumber/api/features/policy_load_errors.feature:103]: @egvili  - should all these be updated too? Failed to load policy....

**Review by orenbm — DISMISSED** [review]

**Review by shulifink — COMMENTED** [review]
@egvili  - I can;t find the change

**Comment by shulifink** [comment]
@egvili  - I got lost... can't find the original message

File: cucumber/api/features/policy_load_errors.feature (hunk 6 above)
Comment (egvili) [reply cucumber/api/features/policy_load_errors.feature:103]: You're right. I will change them as well

**Comment by egvili** [comment]
@shulifink , it's because your original message was resolved. You can see it if you click on `show resolved` next to your comment (one tiny change)

File: app/models/audit/event/policy.rb [hunk 7]
```diff
@@ -38,23 +40,34 @@ def to_s
       # TODO: See issue https://github.com/cyberark/conjur/issues/1608
       # :reek:NilCheck
       def message
-        past_tense_verb = @operation.to_s.chomp('e') + "ed"
-        "#{user&.id} #{past_tense_verb} #{@subject}"
+        attempted_action.message(
+          success_msg: success_message,
+          failure_msg: "Failed to load policy",
```
Comment (shulifink) [inline app/models/audit/event/policy.rb:45]: fine

File: cucumber/api/features/policy_load_errors.feature [hunk 8]
```diff
@@ -30,7 +30,13 @@ Feature: Policy loading error messages
       }
     }
     """
-
+    And there is an audit record matching:
+    """
+      <85>1 * * conjur * policy
+      [auth@43868 user="cucumber:user:admin"][subject@43868]
+      [action@43868 result="failure" operation="add"]
+      [meta sequenceId="4"] Failed to load policy: User 'bob' not found in account 'cucumber'
```
Comment (shulifink) [inline cucumber/api/features/policy_load_errors.feature:38]: fine

File: cucumber/api/features/policy_load_errors.feature [hunk 9]
```diff
@@ -82,3 +95,10 @@ Feature: Policy loading error messages
       }
     }
     """
+    And there is an audit record matching:
+    """
+      <85>1 * * conjur * policy
+      [auth@43868 user="cucumber:user:admin"][subject@43868]
+      [action@43868 result="failure" operation="add"]
+      [meta sequenceId="4"] Failed to load policy: policy_text is not present
```
Comment (shulifink) [inline cucumber/api/features/policy_load_errors.feature:103]:
```suggestion
      [meta sequenceId="4"] Failed to load policy: policy_text not found
```

File: cucumber/api/features/policy_load_errors.feature [hunk 10]
```diff
@@ -61,6 +67,13 @@ Feature: Policy loading error messages
       }
     }
     """
+    And there is an audit record matching:
+    """
+      <85>1 * * conjur * policy
+      [auth@43868 user="cucumber:user:admin"][subject@43868]
+      [action@43868 result="failure" operation="add"]
+      [meta sequenceId="4"] Failed to load policy: policy_text resource has a blank id
```
Comment (shulifink) [inline cucumber/api/features/policy_load_errors.feature:75]:
```suggestion
      [meta sequenceId="4"] Failed to load policy: policy_text resource is empty (has a blank ID)
```

**Review by shulifink — APPROVED** [review]
a couple of small changes in the load_errors file

File: cucumber/api/features/policy_load_errors.feature (hunk 9 above)
Comment (egvili) [reply cucumber/api/features/policy_load_errors.feature:103]: This message has to match to a message that comes from other parties, so it cannot be modified

File: cucumber/api/features/policy_load_errors.feature (hunk 10 above)
Comment (egvili) [reply cucumber/api/features/policy_load_errors.feature:75]: This message has to match to a message that comes from other parties, so it cannot be modified

**Review by orenbm — APPROVED** [review]

**Pull request merged**